odl/operator/fspace_ops.py

```diff
@@ -0,0 +1,412 @@
+                                   edge_order=self.edge_order,
+                                   zero_padding=self.zero_padding)
+
+
+# noinspection PyAbstractClass
+class DiscreteDivergence(Operator):
+    """Divergence operator for any number of dimensions.
+
+    Calls helper function `discrete_part_deriv` for each component of the
+    input product space vector vector. For the adjoint of the
+    `DiscreteDivergence` operator to match the negative `DiscreteGradient`
+    operator requires `zero_padding`.
+    """
+
+    def __init__(self, space, voxel_size=(1,), edge_order=2,
+                 zero_padding=True):
+        """Initialize a `DiscreteDivergence` operator instance.
+
+        Zero padding is required for the negative `DiscreteGradient`
+        operator to match the adjoint of the `DiscreteDivergence` operator.
+
+        Parameters
+        ----------
+        space : `FunctionSpace`
+            The space of elements which the operator is acting on
+        voxel_size : n-tuple of `floats`
+            n-tuple of scalars specifying the sample distances in each
+            dimension. Default distance is 1 in each direction
+        edge_order : {1, 2}, optional
+            Partial derivative is calculated using Nth order accurate
+            differences at the boundaries. Default edge order: 2
+        zero_padding : `bool`, optional
+            Implicit zero padding. Assumes values outside of `space` to be
+            zero. Default: True
+        """
+
+        self.space = space
+        self.voxel_size = voxel_size
+        self.edge_order = edge_order
+        self.zero_padding = zero_padding
+        super().__init__(domain=ProductSpace(space, len(voxel_size)),
+                         range=space, linear=True)
+
+    def _apply(self, rhs, out):
+        """Apply `DiscreteDivergence` operator to `rhs` and store result in
+        `out`.
+
+        Parameters
+        ----------
+        rhs : ``domain`` element
+            `ProductSpaceVector` to which the `DiscreteDivergence` operator is
+            applied to
+        out : ``range`` element
+            Output vector to which the result is written
+
+        Examples
+        --------
+        >>> from odl import uniform_discr, FunctionSpace, IntervalProd
+        >>> def ndvolume(vol_size, ndim, dtype=None):
+        ...     s = [1]
+        ...     vol = np.arange(vol_size, dtype=dtype)
+        ...     for _ in range(ndim - 1):
+        ...         s.insert(0, vol_size)
+        ...         vol = vol * vol.reshape(s)
+        ...     return vol
+        >>> ndim = 4
+        >>> vsize = 6
+        >>> disc = uniform_discr(FunctionSpace(IntervalProd(
+        ... [0.]*ndim, [vsize]*ndim)), [vsize]*ndim)
+        >>> f = disc.element(ndvolume(vsize, ndim, np.int32))
+        >>> B = DiscreteDivergence(disc, (1.,)*ndim, zero_padding=True)
+        >>> g = B.domain.one()
+        >>> Bg = B(g)
+        >>> Bdf = B.adjoint(f)
```

Comment:
Doctests should show what it does, Id prefer actual numbers showing. This is more fit for the actual tests.